File: src-tauri/src/presentation/ipc/features/install/events.rs
use std::sync::Arc;

use tauri::{AppHandle, Emitter};

use super::dtos::{InstallAddEventArgsDto, InstallRemoveEventArgsDto, InstallUpdateEventArgsDto};
use crate::application::{
    event::EventHandler,
    services::install::{InstallAddEventArgs, InstallRemoveEventArgs, InstallUpdateEventArgs},
};

pub struct InstallAddEventEmitter {
    app: AppHandle,
}

pub struct InstallUpdateEventEmitter {
    app: AppHandle,
}

pub struct InstallRemoveEventEmitter {
    app: AppHandle,
}

impl InstallAddEventEmitter {
    pub fn new(app: AppHandle) -> Self {
        Self { app }
    }
}

impl InstallUpdateEventEmitter {
    pub fn new(app: AppHandle) -> Self {
        Self { app }
    }
}

impl InstallRemoveEventEmitter {
    pub fn new(app: AppHandle) -> Self {
        Self { app }
    }
}

impl EventHandler<InstallAddEventArgs> for InstallAddEventEmitter {
    fn invoke(&self, args: Arc<InstallAddEventArgs>) {
        let _ = self
            .app
            .emit("installs::add", &InstallAddEventArgsDto::from(args));
    }
}

impl EventHandler<InstallUpdateEventArgs> for InstallUpdateEventEmitter {
    fn invoke(&self, args: Arc<InstallUpdateEventArgs>) {
        let _ = self
            .app
            .emit("installs::update", &InstallUpdateEventArgsDto::from(args));
    }
}

impl EventHandler<InstallRemoveEventArgs> for InstallRemoveEventEmitter {
    fn invoke(&self, args: Arc<InstallRemoveEventArgs>) {
        let _ = self
            .app
            .emit("installs::remove", &InstallRemoveEventArgsDto::from(args));
    }
}

File: src/lib/ipc/features/install/events.ts
import { listen } from "@tauri-apps/api/event";
import {
  InstallAddEventArgs,
  InstallRemoveEventArgs,
  InstallUpdateEventArgs,
} from "./types";

export const addEvent = {
  subscribe: (handler: (args: InstallAddEventArgs) => void): (() => void) => {
    const unlisten = listen<InstallAddEventArgs>("installs::add", (event) => {
      handler(event.payload);
    });

    return () => {
      unlisten.then((fn) => fn());
    };
  },
};

export const updateEvent = {
  subscribe: (
    handler: (args: InstallUpdateEventArgs) => void,
  ): (() => void) => {
    const unlisten = listen<InstallUpdateEventArgs>(
      "installs::update",
      (event) => {
        handler(event.payload);
      },
    );

    return () => {
      unlisten.then((fn) => fn());
    };
  },
};

export const removeEvent = {
  subscribe: (
    handler: (args: InstallRemoveEventArgs) => void,
  ): (() => void) => {
    const unlisten = listen<InstallRemoveEventArgs>(
      "installs::remove",
      (event) => {
        handler(event.payload);
      },
    );

    return () => {
      unlisten.then((fn) => fn());
    };
  },
};
